src/julia.h

```diff
@@ -928,7 +935,7 @@ STATIC_INLINE int jl_is_datatype_singleton(jl_datatype_t *d)
 STATIC_INLINE int jl_is_datatype_make_singleton(jl_datatype_t *d)
 {
     return (!d->abstract && jl_datatype_size(d) == 0 && d != jl_sym_type && d->name != jl_array_typename &&
-            d->uid != 0 && (d->name->names == jl_emptysvec || !d->mutabl));
+            d->uid != 0 && (jl_datatype_nfields(d) == 0 || !d->mutabl));
```

Comment:
`jl_datatype_nfields` needs to examine `layout`, so it's less likely to be valid to use this predicate after this change (wasn't really valid before anyways though). Lets sink it into jltypes.c so that it isn't exported anymore.